hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/kafka/service/KafkaSingleMessageReader.java

```diff
@@ -1,72 +1,85 @@
+import pl.allegro.tech.hermes.common.message.wrapper.AvroMessageContentWrapper;
 import pl.allegro.tech.hermes.management.domain.topic.SingleMessageReader;
 import pl.allegro.tech.hermes.management.domain.topic.SingleMessageReaderException;
 
 import java.nio.ByteBuffer;
 import java.nio.charset.Charset;
 
+import static java.lang.String.format;
+
 public class KafkaSingleMessageReader implements SingleMessageReader {
 
     private static final Logger logger = LoggerFactory.getLogger(KafkaSingleMessageReader.class);
 
     private final SimpleConsumerPool simpleConsumerPool;
+    private final AvroMessageContentWrapper avroMessageContentWrapper;
 
-    public KafkaSingleMessageReader(SimpleConsumerPool simpleConsumerPool) {
+    public KafkaSingleMessageReader(SimpleConsumerPool simpleConsumerPool, AvroMessageContentWrapper avroMessageContentWrapper) {
         this.simpleConsumerPool = simpleConsumerPool;
+        this.avroMessageContentWrapper = avroMessageContentWrapper;
     }
 
     @Override
-    public String readMessage(TopicName topic, int partition, long offset) {
-        SimpleConsumer simpleConsumer = simpleConsumerPool.get(topic.qualifiedName(), partition);
-
-        FetchResponse fetchResponse = simpleConsumer.fetch(
-            createFetchRequest(simpleConsumer.clientId(), topic, partition, offset, simpleConsumerPool.getBufferSize())
-        );
-
-        String topicName = topic.qualifiedName();
-        if (fetchResponse.hasError()) {
-            logger.error("Cannot read offset {} from topic/partition {}/{}. Error code: {}",
-                    offset, topic.qualifiedName(), partition, fetchResponse.errorCode(topicName, partition));
-            throw new SingleMessageReaderException(String.format("Cannot read offset %d from topic/partition %s/%d. Error code: %d",
-                    offset, topic.qualifiedName(), partition, fetchResponse.errorCode(topicName, partition)));
+    public String readMessage(Topic topic, int partition, long offset) {
+        byte[] bytes = readMessage(topic.getQualifiedName(), partition, offset);
+        if (topic.getContentType() == Topic.ContentType.AVRO) {
+            bytes = convertAvroToJson(topic.getMessageSchema(), bytes);
         }
+        return new String(bytes, Charset.forName("UTF-8"));
+    }
 
-        for (MessageAndOffset messageAndOffset : fetchResponse.messageSet(topicName, partition)) {
-            long currentOffset = messageAndOffset.offset();
-            if (currentOffset < offset) {
-                logger.info("Found an old offset: {} Expecting: {}", currentOffset, offset);
-                continue;
-            }
-            ByteBuffer payload = messageAndOffset.message().payload();
+    private byte[] convertAvroToJson(String schema, byte[] bytes) {
+        return new AvroToJsonConverter(new Schema.Parser().parse(schema))
+                .convert(avroMessageContentWrapper.unwrapContent(bytes).getContent());
```

Comment:
Good question. Maybe always returning wrapped messages is a right call? Right now I cannot convert AVRO to JSON without some AVRO schema containing metadata though...